src/main/java/net/tridentsdk/chat/ChatComponent.java

```diff
@@ -0,0 +1,482 @@
+import lombok.*;
+import lombok.experimental.Accessors;
+
+import java.util.Collections;
+import java.util.List;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.locks.ReadWriteLock;
+import java.util.concurrent.locks.ReentrantReadWriteLock;
+
+/**
+ * Represents a component in a Minecraft chat format.
+ */
+@Accessors(chain = true)
+@NoArgsConstructor
+@AllArgsConstructor(access = AccessLevel.PRIVATE)
+public class ChatComponent {
+
+    @Getter @Setter
+    private String text, translate, selector, insertion, scoreUsername, scoreObjective;
+
+    @Getter @Setter
+    private ChatColor color;
+
+    @Getter @Setter
+    private ClickEvent clickEvent;
+
+    @Getter @Setter
+    private HoverEvent hoverEvent;
+
+    private List<JsonElement> with = Lists.newCopyOnWriteArrayList();
+    private List<ChatComponent> extra = Lists.newCopyOnWriteArrayList();
+    private AtomicBoolean bold, italic, underlined, strikethrough, obfuscated;
+
+    private ReadWriteLock $lock = new ReentrantReadWriteLock();
+
+    /**
+     * Gets all elements attached to the 'with' array.
+     *
+     * The 'with' array is used in conjunction with the 'translate' component.
+     *
+     * @return The with elements.
+     */
+    public List<JsonElement> getWith() {
+        return Collections.unmodifiableList(with);
+    }
+
+    /**
+     * Adds an element to the 'with' array.
+     *
+     * @param element The JSON element.
+     *
+     * @return This object.
+     */
+    public ChatComponent addWith(JsonElement element) {
+        $lock.writeLock().lock();
+        with.add(element);
+        $lock.writeLock().unlock();
+        return this;
+    }
+
+    /**
+     * Adds an string to the 'with' array.
+     *
+     * @param element The string.
+     *
+     * @return This object.
+     */
+    public ChatComponent addWith(String element) {
+        return addWith(new JsonPrimitive(element));
+    }
+
+    /**
+     * Gets all components attached to the 'extra' array.
+     *
+     * Components in this array are appended to this component.
+     *
+     * @return The extra components.
+     */
+    public List<ChatComponent> getExtra() {
+        return Collections.unmodifiableList(extra);
+    }
+
+    /**
+     * Adds a component to the 'extra' array.
+     *
+     * @param component The component.
+     *
+     * @return This component.
+     */
+    public ChatComponent addExtra(ChatComponent component) {
+        if (!hasExtra(component, true)) {
+            $lock.writeLock().lock();
+            this.extra.add(component);
+            $lock.writeLock().unlock();
+        }
+        return this;
+    }
+
+    /**
+     * Adds a string to the 'extra' array.
+     *
+     * @param string The string.
+     *
+     * @return This component.
+     */
+    public ChatComponent addExtra(String string) {
+        $lock.writeLock().lock();
+        this.extra.add(ChatComponent.create().setText(string));
```

Comment:
THREAD-SAFETY
You need to make up your mind here (lol), use thread safe collection OR use the lock.

Please also use the try-finally idiom when using explicit locks as well.